Sources/OpenCombine/Future.swift

```diff
@@ -0,0 +1,119 @@
+//
+//  Future.swift
+//  
+//
+//  Created by Max Desiatov on 24/11/2019.
+//
+
+import COpenCombineHelpers
+
+/// A publisher that eventually produces one value and then finishes or fails.
+public final class Future<Output, Failure>: Publisher where Failure: Error {
+
+    public typealias Promise = (Result<Output, Failure>) -> Void
+
+    private let _lock = UnfairRecursiveLock.allocate()
+    private var _subscriptions: [Conduit] = []
+
+    private var result: Result<Output, Failure>?
+
+    public init(
+        _ attemptToFulfill: @escaping (@escaping Promise) -> Void
+    ) {
+        attemptToFulfill { result in
+            self._lock.do {
+                self.result = result
+                self._publish(result)
+            }
+        }
+    }
+
+    deinit {
+        _lock.deallocate()
+    }
+
+    /// This function is called to attach the specified `Subscriber` to this
+    /// `Publisher` by `subscribe(_:)`
+    ///
+    /// - SeeAlso: `subscribe(_:)`
+    /// - Parameters:
+    ///     - subscriber: The subscriber to attach to this `Publisher`.
+    ///                   once attached it can begin to receive values.
+    public func receive<Downstream: Subscriber>(
+        subscriber: Downstream
+    ) where Output == Downstream.Input, Failure == Downstream.Failure {
+        let subscription = Conduit(parent: self,
+                                   downstream: AnySubscriber(subscriber))
+
+        _subscriptions.append(subscription)
+       subscriber.receive(subscription: subscription)
+    }
+
+    private func _acknowledgeDownstreamDemand() {
+        _lock.do {
+            guard let result = result else { return }
+            _publish(result)
+        }
+    }
+
+    private func _publish(_ result: Result<Output, Failure>) {
+        for subscription in self._subscriptions where !subscription._isCompleted {
+            switch result {
+            case let .success(output) where subscription._demand > 0:
+               subscription._demand += subscription._downstream?.receive(output) ?? .none
+               subscription._demand -= 1
```

Comment:
I think these two lines should be swapped. This is how we usually do it, and in the general case it makes sense if calling the `receive` method causes recursion (in which case the decrementing line is only reached after the recursion is terminated, which is bad).

By the way, recursion. This code:

```swift
var promise: Future<Int, Never>.Promise?
let future = Future<Int, Never> {
    promise = $0
}

let cancellable = future.sink {
    promise?(.success($0 + 1))
    print($0)
}

promise?(.success(0))

cancellable.cancel()
```
should not cause infinite recursion (it should just print 0).

After a brief look at your implementation I'm not sure there won't be infinite recursion in this case. Can you please add a test for this behavior? It'd be good to test that there's no infinite recursion not only when receiving value, but also when we're receiving completion.

Reply:
Great shout, there is indeed recursion in that case! Swapping the lines didn't help though. By the way, as I mentioned, most of this code is a copy with minor edits from `PassthroughSubject`, which has this:

```swift
let newDemand = subscription._downstream?.receive(input) ?? .none
subscription._demand += newDemand
subscription._demand -= 1
```

Reply:
>Swapping the lines didn't help though.

I never said it should have :)

>By the way, as I mentioned, most of this code is a copy with minor edits from `PassthroughSubject`

The `PassthroughSubject` and `CurrentValueSubject` publishers were one of the first things I implemented in OpenCombine, before I gained any experience with it, so they are in many ways not optimal (and maybe even not totally correct).

I'm working on a better implementation in the [`experiment/subscriber-list `](https://github.com/broadwaylamb/OpenCombine/tree/experiment/subscriber-list) branch. It's not of the highest priority though, and I want #85 to be merged first.